Detectors/Raw/include/DetectorsRaw/RawFileWriter.h

```diff
@@ -49,28 +44,49 @@ class RawFileWriter
                                               std::vector<char>& trailer, std::vector<char>& header)>;
   using EmptyPageCallBack = std::function<void(const RDH& rdh, std::vector<char>& emptyHBF)>;
 
-  //=====================================================================================
+  ///=====================================================================================
+  /// output file handler with its own lock
+  struct OutputFile {
+    FILE* handler = nullptr;
+    std::mutex fileMtx;
+    OutputFile() = default;
+    OutputFile(const OutputFile& src) : handler(src.handler) {}
+    OutputFile& operator=(const OutputFile& src)
+    {
+      if (this != &src) {
+        handler = src.handler;
+      }
+      return *this;
+    }
+    void write(const char* data, size_t size);
+  };
+
+  ///=====================================================================================
+  /// Single GBT link helper
   struct LinkData {
     static constexpr int MarginToFlush = 2 * sizeof(RDH); // flush superpage if free space left <= this margin
     RDH rdhCopy;                                          // RDH with the running info of the last RDH seen
     IR updateIR;                                          // IR at which new HBF needs to be created
-    FILE* file = nullptr;                                 // file handler associated with this link
     int lastRDHoffset = -1;                               // position of last RDH in the link buffer
     bool startOfRun = true;                               // to signal if this is the 1st HBF of the run or not
     uint8_t packetCounter = 0;                            // running counter
     uint16_t pageCnt = 0;                                 // running counter
     LinkSubSpec_t subspec = 0;                            // subspec according to DataDistribution
+    int counter = 0;                                      //RSREM
     //
     size_t nTFWritten = 0;    // number of TFs written
     size_t nRDHWritten = 0;   // number of RDHs written
     size_t nBytesWritten = 0; // number of bytes written
     //
     std::string fileName{};                // file name associated with this link
     std::vector<char> buffer;              // buffer to accumulate superpage data
-    const RawFileWriter* writer = nullptr; // pointer on the parent writer
+    RawFileWriter* writer = nullptr;       // pointer on the parent writer
+    std::mutex mtx;
```

Comment:
are you using this mutex anywhere? You probably only need it if used across different function calls.